torch/nested/_internal/ops.py

```diff
@@ -130,22 +137,59 @@ def jagged_unary_pointwise(func, *args, **kwargs):
 
 def jagged_binary_pointwise(func, *args, **kwargs):
     a, b = args[0], args[1]
-    assert isinstance(a, NestedTensor)
-    if isinstance(b, NestedTensor):
-        check_ragged_dim_same(func, args[0], "lhs", args[1], "rhs")
-        b = args[1]._values
-    else:
-        # TODO: Verify this more and handle the a.dim() == b.dim() case specially if needed
-        if a.dim() <= b.dim():
-            # need to use offsets to broadcast across batch dim properly
-            # NB: inefficient fallback here; Triton codegen can help this
-            assert a.shape[0] == b.shape[0]
-            outputs = []
-            for a_comp, b_comp in zip(a.unbind(), b.unbind()):
-                outputs.append(func(a_comp, b_comp, **kwargs))
-            new_values = torch.cat(outputs, dim=0)
-            return NestedTensor(new_values, **extract_kwargs(a))
-    return NestedTensor(func(a._values, b, **kwargs), **extract_kwargs(a))
+    assert isinstance(a, NestedTensor) or isinstance(b, NestedTensor)
+
+    mismatch_error_msg = (
+        f"cannot call binary pointwise function {func.__name__} with inputs of shapes "
+        f"{a.shape} and {b.shape}"
+    )
+
+    # a is NT, b is NT
+    if isinstance(a, NestedTensor) and isinstance(b, NestedTensor):
+        # ex: (B, j0, D) + (B, j0, D)
+        # ex: (B, j0, D) + (B, j0, 1)
+        if raggedness_matches(a, b.shape):
+            return NestedTensor(func(a._values, b._values, *args[2:], **kwargs),
+                                **extract_kwargs(a))
+        raise RuntimeError(mismatch_error_msg)
+
+    # either a is NT or b is NT at this point
+    a_is_nt = isinstance(a, NestedTensor)
+    extracted_kwargs = extract_kwargs(a) if a_is_nt else extract_kwargs(b)
+
+    # === Handle broadcasting across the batch / ragged dims ===
```

Comment:
no action needed, but eventually we'd need autograd support for implicit broadcasting as HSTU folks have found out